lib/mini_defender/detectors/public_domain_detector.rb

```diff
@@ -0,0 +1,130 @@
+# frozen_string_literal: true
+
+module MiniDefender
+  module Detectors
+    class PublicDomainDetector
+      DOT = '.'
+
+      class << self
+        def public_domain?(host)
+          return false unless host
+
+          instance.valid?(host)
+        end
+
+        def instance
+          @instance ||= new
+        end
+      end
+
+      def initialize
+        @tlds = Set.new
+        @special_suffixes = Set.new
+        @local_patterns = Set.new
+        load_patterns
+      end
+
+      def valid?(host)
+        return false unless host && valid_format?(host)
+
+        host = host.to_s.downcase.strip
+
+        return false if local_network?(host)
+
+        parts = host.split(DOT)
+        return false if parts.length < 2 # must have at least SLD + TLD; a domain composed by a TLD, SLD and TRD
+
+        return true if valid_special_suffix?(parts) # special suffixes first (e.g., co.uk or herokuapp.com)
+
+        # for normal domains, validate TLD and ensure SLD exists
+        tld = parts.last
```

Comment:
Will this section of code be able to validate something like `api.business.com.sa`